packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,507 @@
+import 'package:flutter/rendering.dart';
+
+/// A [Widget] that abstracts over a Material slice and a Material gap.
+abstract class MergeableMaterialItem extends StatelessWidget {
+  MergeableMaterialItem({Key key}) : super(key: key);
+
+  /// Return whether the current item is a [MaterialGap].
+  bool get isGap;
+}
+
+/// A [Widget] that can be used as a child to [MergeableMaterial]. It is a slice
+/// of [Material] that animated merges within the list.
+class MaterialSlice extends MergeableMaterialItem {
+  /// Creates a slice of [Material] that's mergeable within a
+  /// [MergeableMaterial].
+  MaterialSlice({
+    Key key,
+    this.child
+  }) : super(key: key);
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  bool get isGap => false;
+
+  @override
+  Widget build(BuildContext context) {
+    return new Material(
+      type: MaterialType.transparency,
+      child: child
+    );
+  }
+}
+
+/// A [Widget] that represents a gap within [MergeableMaterial].
+class MaterialGap extends MergeableMaterialItem {
+  /// Creates a Material gap with a given size.
+  MaterialGap({
+    Key key,
+    this.size: 16.0
+  }) : super(key: key);
+
+  /// The size of the gap, irrespective of the axis used.
+  final double size;
+
+  @override
+  bool get isGap => true;
+
+  // Dummy object; real Widget generated later.
+  @override
+  Widget build(BuildContext context) => null;
+}
+
+/// A [Widget] useful for animating the merging and separating of [Material]
+/// slices. The widget takes [MergeableMaterialItem] children, uses a block
+/// layout algorithm, and separates [MaterialSlice] wherever it sees a
+/// [MaterialGap]. Animations work by figuring out where new gaps were
+/// added/removed when the widget gets rebuilt.
+///
+/// The addition of extra slices is not animated and neither is the adding of
+/// any extra items at the end of the children list.
+class MergeableMaterial extends StatefulWidget {
+  /// Creates a mergeable Material list of items.
+  MergeableMaterial({
+    Key key,
+    this.mainAxis: Axis.vertical,
+    this.elevation: 2,
+    this.children: const <MergeableMaterialItem>[]
+  }) : super(key: key);
+
+  /// The children of the [MergeableMaterial].
+  final List<MergeableMaterialItem> children;
+
+  /// The main layout axis.
+  final Axis mainAxis;
+
+  /// The elevation of all the [Material] slices.
+  final int elevation;
+
+  @override
+  _MergeableMaterialState createState() {
+    return new _MergeableMaterialState(children);
+  }
+}
+
+class _MergeableMaterialState extends State<MergeableMaterial> {
+  _MergeableMaterialState(this.children);
+
+  List<MergeableMaterialItem> children;
+
+  // These lists have to sparse since not all items have all types of
+  // animations. The reason for using List instead of HashMap is because when
+  // adding or removing an item from the children list all keys from the
+  // animation hash maps would need to be updated.
```

Comment:
Maybe have a single list which takes an object for each entry that has the relevant information on it?